scheduler: add metric for pods scheduled after flush

mrvarmazyar opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This PR adds a counter metric to track pods that successfully schedule after being flushed from the unschedulablePods queue due to timeout. When pods sit in unschedulablePods for more than 5 minutes, they're automatically flushed back to activeQ. If these pods schedule shortly after, it may indicate missing queue hint optimizations or event handling bugs.

The metric `scheduler_pod_scheduled_after_flush_total` (ALPHA) increments whenever a pod that was flushed schedules successfully, helping maintainers detect and investigate potential queue hint issues.

#### Which issue(s) this PR is related to:

Fixes #134999

#### Special notes for your reviewer:

This is a minimal change (~17 lines across 3 files):
- Added `WasFlushedFromUnschedulable` boolean field to `QueuedPodInfo`
- Set to true when `flushUnschedulablePodsLeftover()` moves pods due to timeout
- Cleared when pod returns to queue after scheduling failure or when moved by events (not timeout)
- Created counter metric with ALPHA stability
- Increment metric when pods with flag set successfully schedule

The boolean flag approach ensures we only count pods that schedule immediately after timeout flush, avoiding false positives when pods are later moved by cluster events.

#### Does this PR introduce a user-facing change?

```release-note
Added ALPHA metric `scheduler_pod_scheduled_after_flush_total` to count pods scheduled after being flushed from unschedulablePods due to timeout
```

mrvarmazyar committed:
scheduler: add metric for pods scheduled after flush

This metric tracks pods that successfully schedule after being
flushed from unschedulablePods due to timeout. High values may
indicate missing queue hint optimizations or event handling issues.

mrvarmazyar committed:
scheduler: add metric for pods scheduled after flush

Add counter metric to track pods that schedule immediately after
being flushed from unschedulablePods due to timeout. Uses a boolean
flag that is cleared when pods return to queue or move via events.

sanposhiho commented on pkg/scheduler/backend/queue/scheduling_queue.go:
```diff
@@ -1235,6 +1239,13 @@ func (p *PriorityQueue) movePodsToActiveOrBackoffQueue(logger klog.Logger, podIn
+		if event != framework.EventUnschedulableTimeout {
+			pInfo.WasFlushedFromUnschedulable = false
+		}
+
```
Can we just do this at AddUnschedulableIfNotPresent? any reason not to that?

mrvarmazyar committed:
Removing the reduntant WasFlushedFromUnschedulable

macsko commented on pkg/scheduler/schedule_one.go:
```diff
@@ -341,6 +341,10 @@ func (sched *Scheduler) bindingCycle(
 	}
+	// Count pods scheduled after being flushed from unschedulablePods
+	if assumedPodInfo.WasFlushedFromUnschedulable {
+		metrics.PodScheduledAfterFlush.Inc()
```
I think we could call this metric in SchedulingQueue.Done() method instead

mrvarmazyar replied:
Thanks for the suggestion @macsko! I looked into moving the metric to `Done()`, but I see a few challenges:

1. **Done() at line 325 is called before binding completes** - It's called after Permit but before PreBind/Bind
([here](https://github.com/mrvarmazyar/kubernetes/blob/add-pod-flush-metric/pkg/scheduler/schedule_one.go#L325)). Pods can still fail at [PreBind](https://github.com/mrvarmazyar/kubernetes/blob/add-pod-flush-metric/pkg/scheduler/schedule_one.go#L328) or [Bind](https://github.com/mrvarmazyar/kubernetes/blob/add-pod-flush-metric/pkg/scheduler/schedule_one.go#L333) after Done() is called. We'd
  be counting pods that haven't actually scheduled yet.

2. **Done() doesn't have access to QueuedPodInfo** - It only receives `pod UID`, and the inFlightPods map stores `*v1.Pod` not `*QueuedPodInfo`
([here](https://github.com/mrvarmazyar/kubernetes/blob/add-pod-flush-metric/pkg/scheduler/backend/queue/active_queue.go#L322)), so we wouldn't have access to the `WasFlushedFromUnschedulable` flag.

3. **Other Done() calls are in failure paths** - Lines [91](https://github.com/mrvarmazyar/kubernetes/blob/add-pod-flush-metric/pkg/scheduler/schedule_one.go#L91), [96](https://github.com/mrvarmazyar/kubernetes/blob/add-pod-flush-metric/pkg/scheduler/schedule_one.go#L96),[1047](https://github.com/mrvarmazyar/kubernetes/blob/add-pod-flush-metric/pkg/scheduler/schedule_one.go#L1047) are all error/skip cases where we definitely don't want to increment the metric.

Currently the metric is recorded [here](https://github.com/mrvarmazyar/kubernetes/blob/add-pod-flush-metric/pkg/scheduler/schedule_one.go#L345-L347), after the full binding cycle completes successfully
(similar to how `PodSchedulingAttempts` and `PodSchedulingSLIDuration` are recorded [here](https://github.com/mrvarmazyar/kubernetes/blob/add-pod-flush-metric/pkg/scheduler/schedule_one.go#L340-L343)).

Would you suggest modifying Done() to accept additional parameters, or is there another approach you had in mind?

macsko replied:
I was wrong, the actual implementation looks fine

mrvarmazyar committed:
Update pkg/scheduler/metrics/metrics.go

Co-authored-by: Maciej Skoczeń <87243939+macsko@users.noreply.github.com>

mrvarmazyar commented on pkg/scheduler/schedule_one.go:
```diff
@@ -341,6 +341,10 @@ func (sched *Scheduler) bindingCycle(
 	}
+	// Count pods scheduled after being flushed from unschedulablePods
+	if assumedPodInfo.WasFlushedFromUnschedulable {
+		metrics.PodScheduledAfterFlush.Inc()
```
Sure, I'll take a look and will get back to you

mrvarmazyar committed:
Add unit test for WasFlushedFromUnschedulable flag

macsko commented on pkg/scheduler/backend/queue/scheduling_queue_test.go:
```diff
@@ -3023,6 +3023,95 @@ func TestHighPriorityFlushUnschedulablePodsLeftover(t *testing.T) {
+		Status:  v1.ConditionFalse,
+		Reason:  v1.PodReasonUnschedulable,
+		Message: "fake scheduling failure",
+	})
```
Don't need to update the condition

macsko commented on pkg/scheduler/backend/queue/scheduling_queue_test.go:
```diff
@@ -3023,6 +3023,95 @@ func TestHighPriorityFlushUnschedulablePodsLeftover(t *testing.T) {
+	}
+	if pInfo.Pod != pod {
+		t.Errorf("Expected: %v after Pop, but got: %v", pod.Name, pInfo.Pod.Name)
+	}
```
Don't need to check this

macsko commented:
/ok-to-test

mrvarmazyar committed:
test: simplify TestFlushUnschedulablePodsLeftoverSetsFlag
scheduler: add logging for pods scheduled after flush and preserve UnschedulablePlugins

macsko commented on pkg/scheduler/backend/queue/scheduling_queue_test.go:
```diff
@@ -3023,6 +3023,68 @@ func TestHighPriorityFlushUnschedulablePodsLeftover(t *testing.T) {
+
+	internalPInfo := q.unschedulablePods.get(pod)
+	if internalPInfo == nil {
+		t.Fatalf("pod should be in unschedulablePods")
```
```suggestion
		t.Fatalf("Pod should be in unschedulablePods")
```

mrvarmazyar commented on pkg/scheduler/schedule_one.go:
```diff
@@ -341,6 +341,10 @@ func (sched *Scheduler) bindingCycle(
 	}
+	// Count pods scheduled after being flushed from unschedulablePods
+	if assumedPodInfo.WasFlushedFromUnschedulable {
+		metrics.PodScheduledAfterFlush.Inc()
```
Good point! I checked and you're right! `UnschedulablePlugins` is assigned (initially I thought it was inserted):

So it gets completely overwritten on each scheduling attempt. Clearing at `Pop()` is unnecessary because:
- If the pod schedules successfully → we need `UnschedulablePlugins` for logging
- If it fails again → gets overwritten anyway

I've removed the `Clear()` and updated the logging to use `UnschedulablePlugins` directly. Much cleaner approach!
Will push my changes and since I updated my fork, I'll rebase and test the changes

mrvarmazyar committed:
Clear plugins at handleSchedulingFailure and preserve both at Pop

sanposhiho commented on pkg/scheduler/backend/queue/active_queue.go:
```diff
@@ -323,12 +323,13 @@ func (aq *activeQueue) unlockedPop(logger klog.Logger) (*framework.QueuedPodInfo
-	pInfo.UnschedulablePlugins.Clear()
-	pInfo.PendingPlugins.Clear()
 	pInfo.GatingPlugin = ""
 	pInfo.GatingPluginEvents = nil
```
nit: I know GatingPlugin and GatingPluginEvents can still be cleared here as well, but I prefer to clean up all these similar fields at the same place. So, let's create a new method on pInfo, something like `ClearRejectorPlugins()` and clear all of those four fields at `handleSchedulingFailure`?

mrvarmazyar committed:
Refactor plugin clearing to use ClearRejectorPlugins method

mrvarmazyar committed:
Address comments, log level, test assersion consistency and remove unnecessary locks in TestFlushUnschedulablePodsLeftoverSetsFlag

mrvarmazyar commented on pkg/scheduler/backend/queue/scheduling_queue_test.go:
```diff
@@ -3023,6 +3030,68 @@ func TestHighPriorityFlushUnschedulablePodsLeftover(t *testing.T) {
+
+	// Verify flag is cleared when pod returns to queue
+	q.lock.Lock()
+	defer q.lock.Unlock()
```
@macsko Done! All feedback addressed in the latest commit.

sanposhiho reviewed (commented):
/lgtm

Thanks.
Leave /approve for @macsko
